File: zaptorch-service/src/main/java/com/pyamsoft/zaptorch/service/LollipopCamera.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.zaptorch.service;

import android.Manifest;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CameraMetadata;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.os.Build;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.util.Size;
import android.view.Surface;
import com.pyamsoft.pydroid.helper.Checker;
import io.reactivex.Scheduler;
import java.util.ArrayList;
import java.util.List;
import timber.log.Timber;

@TargetApi(Build.VERSION_CODES.LOLLIPOP) class LollipopCamera extends CameraCommon {

  @NonNull private final CameraManager cameraManager;
  @NonNull private final CameraCallback cameraCallback;
  @Nullable private final String flashCameraId;

  LollipopCamera(final @NonNull Context context, final @NonNull VolumeServiceInteractor interactor,
      @NonNull Scheduler obsScheduler, @NonNull Scheduler subScheduler) {
    super(context, interactor, obsScheduler, subScheduler);
    Timber.d("LOLLIPOP CAMERA API");
    this.cameraManager = setupCameraManager(getAppContext());
    this.flashCameraId = setupCamera();
    this.cameraCallback = new CameraCallback(this, cameraManager);
  }

  @CheckResult @NonNull static CameraManager setupCameraManager(final @NonNull Context context) {
    return (CameraManager) context.getApplicationContext().getSystemService(Context.CAMERA_SERVICE);
  }

  @CheckResult @Nullable private String setupCamera() {
    try {
      final String[] cameraList = cameraManager.getCameraIdList();
      for (final String camera : cameraList) {
        final Boolean hasFlash = cameraManager.getCameraCharacteristics(camera)
            .get(CameraCharacteristics.FLASH_INFO_AVAILABLE);
        if (hasFlash != null && hasFlash) {
          return camera;
        }
      }
    } catch (CameraAccessException e) {
      Timber.e(e, "setupCamera ERROR");
    }
    return null;
  }

  @Override public void release() {
    cameraCallback.close();
    super.release();
  }

  @Override public void toggleTorch() {
    if (flashCameraId == null) {
      Timber.e("No setupCamera with Flash");
      startErrorExplanationActivity();
    } else {
      Timber.d("Open setupCamera");
      final int result = cameraCallback.accessCamera(getAppContext(), flashCameraId);
      switch (result) {
        case TYPE_ERROR:
          startErrorExplanationActivity();
          break;
        case TYPE_PERMISSION:
          startPermissionExplanationActivity();
          break;
        default:
          Timber.d("Do nothing");
      }
    }
  }

  @SuppressWarnings("WeakerAccess") static final class CameraCallback
      extends CameraDevice.StateCallback {
    @NonNull final CameraManager manager;
    @NonNull final List<Surface> list;
    @NonNull private final CameraCommon cameraInterface;

    @Nullable CameraDevice cameraDevice;
    @Nullable SessionCallback session;

    @Nullable Size size;
    boolean opened;

    CameraCallback(@NonNull CameraCommon cameraInterface, @NonNull CameraManager manager) {
      this.cameraInterface = Checker.checkNonNull(cameraInterface);
      this.manager = Checker.checkNonNull(manager);
      opened = false;
      list = new ArrayList<>(1);
    }

    @CheckResult @NonNull
    static Size getSmallestSize(final @NonNull CameraManager manager, final @NonNull String id)
        throws CameraAccessException {
      Timber.d("Get stream config map");
      final StreamConfigurationMap map = manager.getCameraCharacteristics(id)
          .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
      if (map == null) {
        throw new IllegalStateException("Camera " + id + "doesn't support any Stream Maps.");
      }

      Timber.d("Get possible output sizes");
      final Size[] outputSizes = map.getOutputSizes(SurfaceTexture.class);
      if (outputSizes == null || outputSizes.length == 0) {
        throw new IllegalStateException("Camera " + id + "doesn't support any outputSize.");
      }

      Timber.d("Select a size");
      Size chosen = outputSizes[0];
      for (final Size s : outputSizes) {
        if (chosen.getWidth() >= s.getWidth() && chosen.getHeight() >= s.getHeight()) {
          chosen = s;
        }
      }
      return chosen;
    }

    void close() {
      // Surface texture is released by CameraManager so we don't have to
      if (opened) {
        if (session != null) {
          Timber.d("close SessionCallback");
          session.close();
          session = null;
        }

        if (cameraDevice != null) {
          Timber.d("close setupCamera device");
          cameraDevice.close();
          cameraDevice = null;
        }

        cameraInterface.notifyCallbackOnClosed();
      }
      opened = false;
    }

    @CheckResult int accessCamera(final @NonNull Context context, final @NonNull String id) {
      if (ActivityCompat.checkSelfPermission(context, Manifest.permission.CAMERA)
          == PackageManager.PERMISSION_GRANTED) {
        try {
          Timber.d("Has setupCamera permission, attempt to access");
          if (opened) {
            Timber.d("Close opened setupCamera");
            close();
          } else {
            Timber.d("Open closed setupCamera");
            manager.openCamera(id, this, null);
          }
          return TYPE_NONE;
        } catch (CameraAccessException e) {
          Timber.e(e, "toggleTorch ERROR");
          return TYPE_ERROR;
        }
      } else {
        Timber.e("Missing setupCamera permission");
        return TYPE_PERMISSION;
      }
    }

    @Override public void onOpened(@NonNull CameraDevice camera) {
      Timber.d("onOpened");
      opened = true;
      this.cameraDevice = camera;

      try {
        Timber.d("create capture builder");
        final CaptureRequest.Builder captureBuilder =
            camera.createCaptureRequest(CameraDevice.TEMPLATE_MANUAL);
        captureBuilder.set(CaptureRequest.CONTROL_AE_MODE, CameraMetadata.CONTROL_AF_MODE_AUTO);
        captureBuilder.set(CaptureRequest.FLASH_MODE, CameraMetadata.FLASH_MODE_TORCH);

        if (size == null) {
          Timber.d("get smallest size for texture");
          size = getSmallestSize(manager, camera.getId());
        } else {
          Timber.d("using cached smallest size");
        }
        // The setupCamera session recycles the surface texture, so we should not have to
        @SuppressLint("Recycle") final SurfaceTexture surfaceTexture = new SurfaceTexture(1);
        surfaceTexture.setDefaultBufferSize(size.getWidth(), size.getHeight());

        Timber.d("add surface to texture");
        final Surface surface = new Surface(surfaceTexture);
        if (list.isEmpty()) {
          list.add(0, surface);
        } else {
          list.set(0, surface);
        }

        Timber.d("add capture target");
        captureBuilder.addTarget(surface);

        Timber.d("create new session callback");
        session = new SessionCallback(captureBuilder.build());

        Timber.d("register capture session");
        camera.createCaptureSession(list, session, null);

        cameraInterface.notifyCallbackOnOpened();
      } catch (CameraAccessException e) {
        Timber.e(e, "onOpened");
      }
    }

    @Override public void onDisconnected(@NonNull CameraDevice cameraDevice) {
      this.cameraDevice = null;
      Timber.d("onDisconnected");
      opened = false;
    }

    @Override public void onError(@NonNull CameraDevice cameraDevice, int i) {
      this.cameraDevice = null;
      Timber.e("onError");
      opened = false;
    }
  }

  @SuppressWarnings("WeakerAccess") static final class SessionCallback
      extends CameraCaptureSession.StateCallback {

    @NonNull final CaptureRequest request;
    @Nullable CameraCaptureSession session;

    SessionCallback(@NonNull CaptureRequest request) {
      this.request = Checker.checkNonNull(request);
    }

    @Override public void onConfigured(@NonNull CameraCaptureSession cameraCaptureSession) {
      Timber.d("Camera configured");
      session = cameraCaptureSession;
      try {
        Timber.d("set repeating");
        cameraCaptureSession.setRepeatingRequest(request, null, null);
      } catch (CameraAccessException e) {
        Timber.e(e, "onConfigured");
      }
    }

    @Override public void onConfigureFailed(@NonNull CameraCaptureSession cameraCaptureSession) {
      Timber.e("onConfigureFailed");
      session = cameraCaptureSession;
      close();
    }

    void close() {
      if (session != null) {
        Timber.d("close session");
        session.close();
        session = null;
      }
    }
  }
}

File: zaptorch-service/src/main/java/com/pyamsoft/zaptorch/service/LollipopCamera.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.zaptorch.service

import android.Manifest
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.SurfaceTexture
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraCaptureSession
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraMetadata
import android.hardware.camera2.CaptureRequest
import android.os.Build
import android.support.annotation.CheckResult
import android.support.v4.content.ContextCompat
import android.util.Size
import android.view.Surface
import com.pyamsoft.zaptorch.api.CameraInterface
import com.pyamsoft.zaptorch.api.VolumeServiceInteractor
import io.reactivex.Scheduler
import timber.log.Timber
import java.util.ArrayList

@TargetApi(Build.VERSION_CODES.LOLLIPOP)
internal class LollipopCamera internal constructor(
    context: Context, interactor: VolumeServiceInteractor, computationScheduler: Scheduler,
    mainScheduler: Scheduler
) :
    CameraCommon(context, interactor, computationScheduler, mainScheduler) {

    private val cameraManager: CameraManager = appContext.getSystemService(
        Context.CAMERA_SERVICE
    ) as CameraManager
    private val flashCameraId = setupCamera()
    private val cameraCallback = CameraCallback(this, cameraManager)

    @CheckResult
    private fun setupCamera(): String? {
        try {
            val cameraList = cameraManager.cameraIdList
            for (camera in cameraList) {
                val hasFlash = cameraManager.getCameraCharacteristics(camera)
                    .get(CameraCharacteristics.FLASH_INFO_AVAILABLE)
                if (hasFlash != null && hasFlash) {
                    return camera
                }
            }
        } catch (e: CameraAccessException) {
            Timber.e(e, "setupCamera ERROR")
        }

        return null
    }

    override fun release() {
        cameraCallback.close()
    }

    override fun toggleTorch() {
        if (flashCameraId == null) {
            Timber.e("No setupCamera with Flash")
            startErrorExplanationActivity()
        } else {
            Timber.d("Open setupCamera")
            val result = cameraCallback.accessCamera(appContext, flashCameraId)
            when (result) {
                CameraInterface.TYPE_ERROR -> startErrorExplanationActivity()
                CameraInterface.TYPE_PERMISSION -> startPermissionExplanationActivity()
                else -> Timber.d("Do nothing")
            }
        }
    }

    internal class CameraCallback internal constructor(
        private val cameraInterface: CameraCommon,
        private val manager: CameraManager
    ) : CameraDevice.StateCallback() {
        private val list: MutableList<Surface>

        private var cameraDevice: CameraDevice? = null
        private var session: SessionCallback? = null

        private var size: Size? = null
        private var opened: Boolean = false

        init {
            opened = false
            list = ArrayList(1)
        }

        fun close() {
            // Surface texture is released by CameraManager so we don't have to
            if (opened) {
                val obj = session
                if (obj != null) {
                    Timber.d("close SessionCallback")
                    obj.close()
                    session = null
                }

                val camera = cameraDevice
                if (camera != null) {
                    Timber.d("close setupCamera device")
                    camera.close()
                    cameraDevice = null
                }

                Timber.d("Release surfaces")
                for (surface in list) {
                    surface.release()
                }
                list.clear()
                cameraInterface.notifyCallbackOnClosed()
            }
            opened = false
        }

        @CheckResult
        fun accessCamera(context: Context, id: String): Int {
            if (ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.CAMERA
                ) == PackageManager.PERMISSION_GRANTED) {
                return try {
                    Timber.d("Has setupCamera permission, attempt to access")
                    if (opened) {
                        Timber.d("Close opened setupCamera")
                        close()
                    } else {
                        Timber.d("Open closed setupCamera")
                        manager.openCamera(id, this, null)
                    }

                    // Return
                    CameraInterface.TYPE_NONE
                } catch (e: CameraAccessException) {
                    Timber.e(e, "toggleTorch ERROR")

                    // Return
                    CameraInterface.TYPE_ERROR
                }
            } else {
                Timber.e("Missing setupCamera permission")
                return CameraInterface.TYPE_PERMISSION
            }
        }

        @SuppressLint("Recycle")
        override fun onOpened(camera: CameraDevice) {
            Timber.d("onOpened")
            opened = true
            this.cameraDevice = camera

            try {
                Timber.d("create capture builder")
                val captureBuilder = camera.createCaptureRequest(CameraDevice.TEMPLATE_MANUAL)
                captureBuilder.set(
                    CaptureRequest.CONTROL_AE_MODE,
                    CameraMetadata.CONTROL_AF_MODE_AUTO
                )
                captureBuilder.set(CaptureRequest.FLASH_MODE, CameraMetadata.FLASH_MODE_TORCH)

                if (size == null) {
                    Timber.d("get smallest size for texture")
                    size = getSmallestSize(manager, camera.id)
                } else {
                    Timber.d("using cached smallest size")
                }

                // The setupCamera session recycles the surface texture, so we should not have to
                val surfaceTexture = SurfaceTexture(1)
                val smallestSize = size ?: throw RuntimeException(
                    "Size is NULL, This should not happen!"
                )
                surfaceTexture.setDefaultBufferSize(smallestSize.width, smallestSize.height)

                Timber.d("add surface to texture")
                val surface = Surface(surfaceTexture)
                if (list.isEmpty()) {
                    list.add(0, surface)
                } else {
                    list[0] = surface
                }

                Timber.d("add capture target")
                captureBuilder.addTarget(surface)

                Timber.d("create new session callback")
                val callback = SessionCallback(captureBuilder.build())
                session = callback

                Timber.d("register capture session")
                camera.createCaptureSession(list, callback, null)

                cameraInterface.notifyCallbackOnOpened()
            } catch (e: CameraAccessException) {
                Timber.e(e, "onOpened")
            }
        }

        override fun onDisconnected(cameraDevice: CameraDevice) {
            this.cameraDevice = null
            Timber.d("onDisconnected")
            opened = false
        }

        override fun onError(cameraDevice: CameraDevice, i: Int) {
            this.cameraDevice = null
            Timber.e("onError")
            opened = false
        }

        companion object {

            @CheckResult
            @JvmStatic
            @Throws(CameraAccessException::class)
            fun getSmallestSize(
                manager: CameraManager, id: String
            ): Size {
                Timber.d("Get stream config map")
                val map = manager.getCameraCharacteristics(id)
                    .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
                        ?: throw IllegalStateException("Camera $id doesn't support any Stream Maps.")

                Timber.d("Get possible output sizes")
                val outputSizes = map.getOutputSizes(SurfaceTexture::class.java)
                if (outputSizes == null || outputSizes.isEmpty()) {
                    throw IllegalStateException("Camera " + id + "doesn't support any outputSize.")
                }

                Timber.d("Select a size")
                var chosen = outputSizes[0]
                outputSizes
                    .asSequence()
                    .filter { chosen.width >= it.width && chosen.height >= it.height }
                    .forEach { chosen = it }
                return chosen
            }
        }
    }

    internal class SessionCallback internal constructor(
        private val request: CaptureRequest
    ) : CameraCaptureSession.StateCallback() {

        private var session: CameraCaptureSession? = null

        override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
            Timber.d("Camera configured")
            session = cameraCaptureSession
            try {
                Timber.d("set repeating")
                cameraCaptureSession.setRepeatingRequest(request, null, null)
            } catch (e: CameraAccessException) {
                Timber.e(e, "onConfigured")
            }
        }

        override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
            Timber.e("onConfigureFailed")
            session = cameraCaptureSession
            close()
        }

        fun close() {
            val obj = session
            if (obj != null) {
                Timber.d("close session")
                obj.close()
                session = null
            }
        }
    }
}
